sql: plumb contexts to logs

RaduBerinde opened the pull request:
Fixes context for log callsites in `sql/*.go`, with the exception of schema
changer and leases.

Part of #7992.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8147)

<!-- Reviewable:end -->

tamird commented:
Reviewed 19 of 19 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[sql/executor.go, line 159 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwpFOXGqCFXDWDRhq2:-KNwpFOXGqCFXDWDRhq3:-1635794839) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/executor.go#L159)):*
> ```Go
> // a Executor; the rest will have sane defaults set if omitted.
> type ExecutorContext struct {
> 	context.Context
> ```

can you give this a name? i'm not seeing the benefit of embedding.

---

*[sql/executor.go, line 306 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwpNbqmBJgyUFFCRlc:-KNwpNbrTskRU7HAsUmq:-968917031) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/executor.go#L306)):*
> ```Go
> ) ([]ResultColumn, error) {
> 	if log.V(2) {
> 		log.Infof(ctx, "preparing statement: %s", query)
> ```

throughout: it's now weird that we're passing a context into this. how do i know if i should use the passed-in one or `e.ctx.Context`?

---

*[sql/executor.go, line 657 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwpVIxlh9e-j-czXNt:-KNwpVIxlh9e-j-czXNu:857927589) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/executor.go#L657)):*
> ```Go
> 		panic(fmt.Sprintf("inconsistent planMaker txn state. txnState: %+v", txnState))
> 	}
> 	ctx := context.Background()
> ```

e.ctx.Context?

---

*[sql/main_test.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwoYynDPcAmZdSfSCe:-KNwoYynDPcAmZdSfSCf:523540022) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/main_test.go#L43)):*
> ```Go
> )
> 
> var testCtx = context.Background()
> ```

nit: i'd prefer to just inline this as we do today

---

*[sql/planner.go, line 152 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwpzcJNXemNPzabCCv:-KNwpzcJNXemNPzabCCw:783323420) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/planner.go#L152)):*
> ```Go
> func (p *planner) ctx() context.Context {
> 	if p == nil {
> 		return context.Background()
> ```

i'd push this to the caller. the planner is often constructed by the executor, we should hook up those contexts.

---

*[sql/trace.go, line 66 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwouaWV9SVaZpD7PRC:-KNwouaWV9SVaZpD7PRD:-563067420) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/trace.go#L66)):*
> ```Go
> 		},
> 		sort: &sortNode{
> 			// Don't use the planner context: we don't want any events from this
> ```

wah, why?

---

*[sql/update.go, line 283 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwp6zFld3f9Gt6MUHd:-KNwp6zFld3f9Gt6MUHe:510325593) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/update.go#L283)):*
> ```Go
> 
> func (u *updateNode) Next() (bool, error) {
> 	ctx := u.p.ctx()
> ```

this is used in two places - i'd inline.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR!

---

Review status: 15 of 19 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[sql/executor.go, line 159 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwpFOXGqCFXDWDRhq2:-KO5Szgg4nuCWU5SPP7v:1318953554) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/executor.go#L159)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you give this a name? i'm not seeing the benefit of embedding.
</details>
Done. The benefit was able to pass `e.ctx` to log functions directly, but there were only a couple of instances.

---

*[sql/executor.go, line 306 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwpNbqmBJgyUFFCRlc:-KO5TT4NzsfnMN4JAEsl:-1692626031) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/executor.go#L306)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> throughout: it's now weird that we're passing a context into this. how do i know if i should use the passed-in one or `e.ctx.Context`?
</details>
`e.ctx.Context` is the "generic" context of the executor (not associated with a particular operation). The passed context can be particular to the operation we are performing.

---

*[sql/executor.go, line 657 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwpVIxlh9e-j-czXNt:-KO5TolQyiXIfKznkFRZ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/executor.go#L657)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> e.ctx.Context?
</details>
Done.

---

*[sql/main_test.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwoYynDPcAmZdSfSCe:-KO5RNWp6PxbgDxzfN-x:1760673619) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/main_test.go#L43)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: i'd prefer to just inline this as we do today
</details>
Why?
I was thinking that in the future we can set up `testCtx` to make it easier to extract just the test messages (e.g. add a `test` "tag").

---

*[sql/planner.go, line 152 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwpzcJNXemNPzabCCv:-KO5TAuAgvtxdRb66yT7:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/planner.go#L152)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i'd push this to the caller. the planner is often constructed by the executor, we should hook up those contexts.
</details>
Done.

---

*[sql/trace.go, line 66 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwouaWV9SVaZpD7PRC:-KNxYVT9DX3YeRjxa77n:118105003) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/trace.go#L66)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> wah, why?
</details>
The sort node is sorting the trace events themselves (when we do EXPLAIN TRACE). If we hook it up to the context, those sort-related events will show up in the EXPLAIN TRACE output.

---

*[sql/update.go, line 283 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwp6zFld3f9Gt6MUHd:-KO5TR-Nn8YPxuMj1njW:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/update.go#L283)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is used in two places - i'd inline.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-:-KO5R7Dnoinl8dKN26UX:-485616692)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r2.
Review status: 19 of 20 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[sql/executor.go, line 306 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwpNbqmBJgyUFFCRlc:-KO5Yu7zxRuL4CHDXBfU:755752002) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/executor.go#L306)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> `e.ctx.Context` is the "generic" context of the executor (not associated with a particular operation). The passed context can be particular to the operation we are performing.
</details>
OK, we need to codify that somewhere. Alternatively, put a getter on `e.ctx` that takes a possibly-nil context.

---

*[sql/main_test.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwoYynDPcAmZdSfSCe:-KO5ZCK8ymNezSJ_mBp9:-848690942) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/main_test.go#L43)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Why?
> I was thinking that in the future we can set up `testCtx` to make it easier to extract just the test messages (e.g. add a `test` "tag").
</details>
at the very least it is usually a bad idea to have a package-level `var`. Can you make this a function instead?

---

*[sql/trace.go, line 66 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwouaWV9SVaZpD7PRC:-KO5ZTpnozwRPuPCfjBR:1926329331) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/trace.go#L66)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> The sort node is sorting the trace events themselves (when we do EXPLAIN TRACE). If we hook it up to the context, those sort-related events will show up in the EXPLAIN TRACE output.
</details>
Please expand the comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-:-KO5ZX-KFFsXUxEMX0UO:1417088721)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 19 of 20 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[sql/executor.go, line 306 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwpNbqmBJgyUFFCRlc:-KO5_YxAWx8BaHknF5K6:79147948) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/executor.go#L306)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> OK, we need to codify that somewhere. Alternatively, put a getter on `e.ctx` that takes a possibly-nil context.
</details>
I don't really understand the suggestion.. Why would we call that getter here, where we know we have a non-nil ctx?

Isn't this the typical "pattern"? You pass a context for whatever you are doing.. There is always a "more generic" context to fall back to, I don't understand why that is confusing.

---

*[sql/main_test.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwoYynDPcAmZdSfSCe:-KO5a5TePZpvw-kAXwgm:1036467884) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/main_test.go#L43)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> at the very least it is usually a bad idea to have a package-level `var`. Can you make this a function instead?
</details>
Ok, reverted to context.Background(). If we want a more specific test context, we have to go through all test logs and fix them anyway.

---

*[sql/trace.go, line 66 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwouaWV9SVaZpD7PRC:-KO5aUpdmGKk-hpWJaEd:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/trace.go#L66)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Please expand the comment.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[server/status_test.go, line 252 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KO5c_ibPaFeSgN92-XR:-KO5c_icJ1Ex6AwagYwP:297554851) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c3f5b6c2f42f829acb8b4408c1309a24446a83e/server/status_test.go#L252)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 	if log.V(3) {
> 		t.Skip("Test only works with low verbosity levels")
> ```

whoa. why? might be good to file an issue

---

*[sql/executor.go, line 306 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwpNbqmBJgyUFFCRlc:-KO5beH6p9Pl0ouY_LGB:220726339) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/executor.go#L306)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> I don't really understand the suggestion.. Why would we call that getter here, where we know we have a non-nil ctx?
> 
> Isn't this the typical "pattern"? You pass a context for whatever you are doing.. There is always a "more generic" context to fall back to, I don't understand why that is confusing.
</details>
Because it's almost never the case that you have both. The "more generic" context is `context.Background`? That's pretty special. What I'm getting at is this pattern is error prone.

As an example, here's what was done in the standard library:
 - passing a context: https://tip.golang.org/pkg/net/#Dialer.DialContext
 - storing a context: https://tip.golang.org/pkg/net/http/#Request.Context

Note that, for instance, [`net.Dialer`](https://tip.golang.org/pkg/net/#Dialer) does not contain a context - the only way to pass one is via `DialContext`. They've been pretty deliberate about avoiding mixing the patterns, and so should we.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[server/status_test.go, line 252 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KO5c_ibPaFeSgN92-XR:-KO5cpt84Q7N4bUlCZMb:845227419) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c3f5b6c2f42f829acb8b4408c1309a24446a83e/server/status_test.go#L252)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> whoa. why? might be good to file an issue
</details>
Well, take a look at what the test does exactly and reason for yourself if it makes sense to spend energy fixing it for higher Vs.

---

*[sql/executor.go, line 306 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwpNbqmBJgyUFFCRlc:-KO5d9XngF3eka3LQOij:-1658910520) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/executor.go#L306)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Because it's almost never the case that you have both. The "more generic" context is `context.Background`? That's pretty special. What I'm getting at is this pattern is error prone.
> 
> As an example, here's what was done in the standard library:
>  - passing a context: https://tip.golang.org/pkg/net/#Dialer.DialContext
>  - storing a context: https://tip.golang.org/pkg/net/http/#Request.Context
> 
> Note that, for instance, [`net.Dialer`](https://tip.golang.org/pkg/net/#Dialer) does not contain a context - the only way to pass one is via `DialContext`. They've been pretty deliberate about avoiding mixing the patterns, and so should we.
</details>
The "more generic" context is the Executor context. This context is associated with the Executor (which is longer lived than a single operation). We are NOT storing a context like the Request.Context example - the Executor context never changes for the life of the Executor.

No matter what you are doing, you are doing it inside a logical hierarchy. Here we run an operation as part of an Executor which is part of a SQL server which is part of a Server. Any level in this hierarchy can (and probably should) have its own context.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[server/status_test.go, line 252 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KO5c_ibPaFeSgN92-XR:-KO5fBso2T224Vm_HGbJ:1459754462) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c3f5b6c2f42f829acb8b4408c1309a24446a83e/server/status_test.go#L252)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Well, take a look at what the test does exactly and reason for yourself if it makes sense to spend energy fixing it for higher Vs.
</details>
I understand the general reason - I was wondering out loud if it wasn't worth mentioning the proximate cause.

---

*[sql/executor.go, line 306 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwpNbqmBJgyUFFCRlc:-KO5fT6r_GA700hbDLwX:2020548490) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/executor.go#L306)):*
> We are NOT storing a context like the Request.Context example - the Executor context never changes for the life of the Executor.

I'm not following you here. In the Request.Context case the context also never changes. The only way to change the context is to use https://tip.golang.org/pkg/net/http/#Request.WithContext, which returns a new request.

> No matter what you are doing, you are doing it inside a logical hierarchy. Here we run an operation as part of an Executor which is part of a SQL server which is part of a Server. Any level in this hierarchy can (and probably should) have its own context.

Yes, I agree with this. What I'm saying is that it is odd to mix the two paradigms (context flows through execution; context is long lived).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[sql/executor.go, line 306 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KNwpNbqmBJgyUFFCRlc:-KO5gYu_j1lCauuo-XqB:1093189160) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2a8819ef5b63eafe99ce5e71d782a6ea15aaf3/sql/executor.go#L306)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> > We are NOT storing a context like the Request.Context example - the Executor context never changes for the life of the Executor.
> 
> I'm not following you here. In the Request.Context case the context also never changes. The only way to change the context is to use https://tip.golang.org/pkg/net/http/#Request.WithContext, which returns a new request.
> 
> > No matter what you are doing, you are doing it inside a logical hierarchy. Here we run an operation as part of an Executor which is part of a SQL server which is part of a Server. Any level in this hierarchy can (and probably should) have its own context.
> 
> Yes, I agree with this. What I'm saying is that it is odd to mix the two paradigms (context flows through execution; context is long lived).
</details>
Ok, this is getting confusing because we are now intermingling two different discussions.

This conversation started with the question of why we have two possible contexts (one associated with the operation, one associated with the Executor). So it looks like that's settled now.

Now we are talking about intermingling two coding patterns. We are using the storing pattern as much as possible (we are storing the context in the txn). These functions are "entry points" in that we may not yet have a txn when we call them. Just like we are passing a ctx to `WithContext`, you have to pass a context until you have the structure you store it in. Granted, our stuff is way more complicated than that; there is probably some refactoring that would make this nicer but it's beyond the scope and intent of this change.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Ok, with Peter's recent change we can now use the Session Context(). Updating the Executor interface to not pass contexts. We can pass a base context to `NewSession` if we have one. These changes are in r6. The current change is r5+r6. I will verify COCKROACH_TRACE_SQL still works as before. @petermattis 

---

Review status: 15 of 22 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-:-KOAmwtgopt9BT06B_3b:-1749295004)*
<!-- Sent from Reviewable.io -->

tamird commented:
I'm gonna hold off until peter's PR goes in to review this.

---

Review status: 15 of 23 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-:-KOAqooHdjM8zvI9P0yi:-1094821827)*
<!-- Sent from Reviewable.io -->

petermattis commented:
My PR is in. 

---

Review status: 15 of 23 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-:-KOAtBkoKhv64GBA8OG0:-512106338)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Skimmed this and it generally looks good. Someone else should give it a more thorough review.

---

Review status: 15 of 23 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/planner.go, line 150 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOAu1BtjfoU-HVFHYOK:-KOAu1Bu670xBNkPd93m:1760303974) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc9d48912522eef31319506627c3acfc05566ee1/sql/planner.go#L150)):*
> ```Go
> // ctx returns a context suitable for logging/tracing - either the txn's context
> // (if there is one), or the executor context.
> func (p *planner) ctx() context.Context {
> ```

Inconsistent to have this method named `ctx` but to have the one in `Session` named `Context`. I prefer `ctx`.

---

*[sql/planner.go, line 154 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOAtrXH5YjeIyEG4r9v:-KOAtrXH5YjeIyEG4r9w:-1196661902) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc9d48912522eef31319506627c3acfc05566ee1/sql/planner.go#L154)):*
> ```Go
> 		return p.txn.Context
> 	}
> 	return p.execCtx.Context
> ```

Could/should this method return `p.session.Context()`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-:-KOAuKQ6DRaTa--7NGSZ:1975762305)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod the context in SessionArgs

also have a slight preference for squashing this into one commit

---

Reviewed 4 of 4 files at r5, 5 of 5 files at r6, 5 of 5 files at r7, 6 of 6 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/session.go, line 68 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOAxpPCIk0WJDpMKjNX:-KOAxpPDmauGAEoekdKz:-1306769686) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2a8ff9abd0b5dc2bded7fd799e7268f9663fbb4/sql/session.go#L68)):*
> ```Go
> 	User     string
> 	// BaseCtx is the context from which the session context is derived. If it
> 	// is nil, the Executor context will be used.
> ```

I have a slight preference for this being a separate argument to NewSession. It looks like there's only really one place (outside of tests) where it is non-nil, and since you have the executor there, it's quite easy to reach in and grab it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-:-KOAywTPedI64vqE-j-W:-645066674)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/planner.go, line 150 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOAu1BtjfoU-HVFHYOK:-KOBD-2BqjYwY6dY8d5U:1812987819) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc9d48912522eef31319506627c3acfc05566ee1/sql/planner.go#L150)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Inconsistent to have this method named `ctx` but to have the one in `Session` named `Context`. I prefer `ctx`.
</details>
It would have to be `Ctx()`, we are using it in `pgwire`. Should I make this one `Ctx()` as well or leave it as `ctx()`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/planner.go, line 150 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOAu1BtjfoU-HVFHYOK:-KOBHXhAiaouJ0phnU6Q:-922897686) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc9d48912522eef31319506627c3acfc05566ee1/sql/planner.go#L150)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> It would have to be `Ctx()`, we are using it in `pgwire`. Should I make this one `Ctx()` as well or leave it as `ctx()`?
</details>
Leave this one as `ctx` and make the session one `Ctx`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 16 of 23 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[sql/planner.go, line 150 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOAu1BtjfoU-HVFHYOK:-KOBIT3W-q7nBD2c6kng:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc9d48912522eef31319506627c3acfc05566ee1/sql/planner.go#L150)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Leave this one as `ctx` and make the session one `Ctx`.
</details>
Done.

---

*[sql/planner.go, line 154 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOAtrXH5YjeIyEG4r9v:-KOBISEFgkenu-RmMikV:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc9d48912522eef31319506627c3acfc05566ee1/sql/planner.go#L154)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Could/should this method return `p.session.Context()`?
</details>
Done.

---

*[sql/session.go, line 68 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOAxpPCIk0WJDpMKjNX:-KOBIRiBUFFNgv3k-ula:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2a8ff9abd0b5dc2bded7fd799e7268f9663fbb4/sql/session.go#L68)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I have a slight preference for this being a separate argument to NewSession. It looks like there's only really one place (outside of tests) where it is non-nil, and since you have the executor there, it's quite easy to reach in and grab it.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 7 of 7 files at r9.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[sql/session.go, line 72 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOBNZLlsWRRlVfu-his:-KOBNZLlsWRRlVfu-hit:-775309698) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd32f195cfa52def4c8cf90cda665d510746ab/sql/session.go#L72)):*
> ```Go
> // ctx can be nil (in which case the Executor's context will be used).
> // remote can be nil.
> func NewSession(args SessionArgs, e *Executor, ctx context.Context, remote net.Addr) *Session {
> ```

sorry to nag, ctx should be the first argument.

---

*[sql/session.go, line 94 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOBOA2BojWTveAVF8OV:-KOBOA2BojWTveAVF8OW:-1525855897) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd32f195cfa52def4c8cf90cda665d510746ab/sql/session.go#L94)):*
> ```Go
> 	s.Trace = trace.New("sql."+args.User, remoteStr)
> 	s.Trace.SetMaxEvents(100)
> 	if ctx == nil {
> ```

i left a comment at the caller, but i'd rather see both nil checks removed.

When is it legal for the executor's context to be nil? if this is about saving some boilerplate in tests, i'd rather we had the boilerplate

---

*[sql/table_test.go, line 257 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOBO5Xyq4r4Fk6rRBx0:-KOBO5XzbPYgP74Hccnu:-434500904) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd32f195cfa52def4c8cf90cda665d510746ab/sql/table_test.go#L257)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 
> 	p := planner{session: &Session{context: context.Background()}}
> ```

how come this was needed?

---

*[sql/pgwire/v3.go, line 136 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOBNkW4PKge2febkDDg:-KOBNkW4PKge2febkDDh:1374204079) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd32f195cfa52def4c8cf90cda665d510746ab/sql/pgwire/v3.go#L136)):*
> ```Go
> 		writeBuf: writeBuffer{bytecount: metrics.bytesOutCount},
> 		metrics:  metrics,
> 		session:  sql.NewSession(sessionArgs, executor, nil, conn.RemoteAddr()),
> ```

this is the only place that passes nil. why not pass the executor's context explicitly?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[sql/session.go, line 72 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOBNZLlsWRRlVfu-his:-KOBPEjmENuCx_hp_5pR:-476451761) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd32f195cfa52def4c8cf90cda665d510746ab/sql/session.go#L72)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> sorry to nag, ctx should be the first argument.
</details>
I put it there because it has this special thing where if it's `nil`, we use the one from the `Executor`, so it felt natural for it to come after the executor arg. If we can change this as you suggest in the other comments, I'll move it up.

---

*[sql/session.go, line 94 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOBOA2BojWTveAVF8OV:-KOBPiCgaMf66U4uwr5N:-905281423) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd32f195cfa52def4c8cf90cda665d510746ab/sql/session.go#L94)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i left a comment at the caller, but i'd rather see both nil checks removed.
> 
> When is it legal for the executor's context to be nil? if this is about saving some boilerplate in tests, i'd rather we had the boilerplate
</details>
Yeah it's only for some tests (v3_test.go), but they're outside of this package and I couldn't set the context. I guess I could export a `sql.DummyExecutor()` for use with these tests?

---

*[sql/table_test.go, line 257 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOBO5Xyq4r4Fk6rRBx0:-KOBQG2NY4iU_GBaLPrQ:-1337669320) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd32f195cfa52def4c8cf90cda665d510746ab/sql/table_test.go#L257)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how come this was needed?
</details>
we do a `planner.ctx()` does `session.Ctx()` now so we need to have a session.

---

*[sql/pgwire/v3.go, line 136 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOBNkW4PKge2febkDDg:-KOBQm2oRiNq4_eDfSP3:-445583374) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd32f195cfa52def4c8cf90cda665d510746ab/sql/pgwire/v3.go#L136)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is the only place that passes nil. why not pass the executor's context explicitly?
</details>
The context isn't available from this package.. Would you prefer to export some stuff from Executor just to do this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[sql/pgwire/v3.go, line 136 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOBNkW4PKge2febkDDg:-KOBS-dsbrGupwObz23r:250187830) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd32f195cfa52def4c8cf90cda665d510746ab/sql/pgwire/v3.go#L136)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> The context isn't available from this package.. Would you prefer to export some stuff from Executor just to do this?
</details>
I can also do two functions `NewSession` and `NewSessionWithCtx` and have the former call the latter with the executor context. Let me know.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[sql/session.go, line 72 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOBNZLlsWRRlVfu-his:-KOBUcJd5myzSmfqWnTO:119445978) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd32f195cfa52def4c8cf90cda665d510746ab/sql/session.go#L72)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> I put it there because it has this special thing where if it's `nil`, we use the one from the `Executor`, so it felt natural for it to come after the executor arg. If we can change this as you suggest in the other comments, I'll move it up.
</details>
Yep, let's do it.

---

*[sql/session.go, line 94 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOBOA2BojWTveAVF8OV:-KOBSzdnArEfFBxbNazi:562670172) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd32f195cfa52def4c8cf90cda665d510746ab/sql/session.go#L94)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Yeah it's only for some tests (v3_test.go), but they're outside of this package and I couldn't set the context. I guess I could export a `sql.DummyExecutor()` for use with these tests?
</details>
Yeah, DummyExecutor seems reasonable, though you may not need to if you export a getter as suggested elsewhere.

---

*[sql/pgwire/v3.go, line 136 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOBNkW4PKge2febkDDg:-KOBShsGyJ3mrqEjM6pp:1305909274) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd32f195cfa52def4c8cf90cda665d510746ab/sql/pgwire/v3.go#L136)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> I can also do two functions `NewSession` and `NewSessionWithCtx` and have the former call the latter with the executor context. Let me know.
</details>
let's export a getter to the executor's context

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[sql/session.go, line 72 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOBNZLlsWRRlVfu-his:-KOBYt9c1404j11W9reg:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd32f195cfa52def4c8cf90cda665d510746ab/sql/session.go#L72)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Yep, let's do it.
</details>
Done.

---

*[sql/session.go, line 94 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOBOA2BojWTveAVF8OV:-KOBYswEwszmH8KTE6Hj:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd32f195cfa52def4c8cf90cda665d510746ab/sql/session.go#L94)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Yeah, DummyExecutor seems reasonable, though you may not need to if you export a getter as suggested elsewhere.
</details>
Done.

---

*[sql/pgwire/v3.go, line 136 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-KOBNkW4PKge2febkDDg:-KOBYtvyTWFOZ0l9PMkG:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/66cd32f195cfa52def4c8cf90cda665d510746ab/sql/pgwire/v3.go#L136)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> let's export a getter to the executor's context
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 6 of 6 files at r10.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8147#-:-KOB_VAdtWLtiEdbYKcZ:-1868698381)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR!

RaduBerinde committed:
sql: plumb contexts to logs

Fixes context for log callsites in `sql/*.go`, with the exception of schema
changer and leases.

We can now pass a "base context" to NewSession. Any Executor calls use the
session's context; removing ctx arguments in the interface, and updating
admin.go accordingly.

Part of #7992.
